### ISSUE TITLE
TypeError: compiler.modifiedFiles is not iterable (5.31.1)

### ISSUE THREAD

**Issue by alan-agius4** [webpack/webpack#13407]
# Bug report

**What is the current behavior?**
It appears that https://github.com/webpack/webpack/pull/13399 introduced a regression where `compiler.modifiedFiles` and `compiler.removedFiles` can now have a value of `undefined`, however based on the API defined in the TypeScript type definition these shouldn't be undefined https://github.com/webpack/webpack/blob/1d2253fdab5825cbe6f25b8075d214af9db4ea18/types.d.ts#L1829-L1830

Thankfully, in the Angular CLI we pin the Webpack version to a specific patch version so users aren't effect by this change, but there might users/plugins that might be effected by this change.

**If the current behavior is a bug, please provide the steps to reproduce.**
The bug is that `compiler.modifiedFiles` and `compiler.removedFiles` can now be `undefined`. 

I can happily provide a reproduction, but in this case I don't think it is needed.

Related CI failure: https://github.com/angular/angular-cli/pull/20844 and stack
```
    TypeError: compiler.modifiedFiles is not iterable
        at packages/ngtools/webpack/src/ivy/plugin.ts:208:48
        at Hook.eval (eval at create (/b/f/w/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:13:1)
        at Compiler.newCompilation (/b/f/w/node_modules/webpack/lib/Compiler.js:1030:30)
        at /b/f/w/node_modules/webpack/lib/Compiler.js:1073:29
        at Hook.eval [as callAsync] (eval at create (/b/f/w/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
        at Compiler.compile (/b/f/w/node_modules/webpack/lib/Compiler.js:1068:28)
        at /b/f/w/node_modules/webpack/lib/Watching.js:184:19
        at Hook.eval [as callAsync] (eval at create (/b/f/w/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
        at run (/b/f/w/node_modules/webpack/lib/Watching.js:138:33)
        at Watching._go (/b/f/w/node_modules/webpack/lib/Watching.js:188:3)
        at /b/f/w/node_modules/webpack/lib/Watching.js:237:12
        at /b/f/w/node_modules/webpack/lib/HookWebpackError.js:69:3
        at Hook.eval [as callAsync] (eval at create (/b/f/w/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
        at Cache.storeBuildDependencies (/b/f/w/node_modules/webpack/lib/Cache.js:122:37)
        at Watching._done (/b/f/w/node_modules/webpack/lib/Watching.js:232:25)
        at onCompiled (/b/f/w/node_modules/webpack/lib/Watching.js:142:36)
        at /b/f/w/node_modules/webpack/lib/Compiler.js:1103:17
```

**What is the expected behavior?**
The the types and actual returned value from the logic match.

Either the types are updated or the logic is updated to retain v5.31.0 behaviour were both of the above mentioned properties can never be `undefined`.

**Other relevant information:**
webpack version:  5.31.1
Node.js version: 12.20.2
Operating System: darwin x64
Additional tools: n/a

**Comment by alexander-akait** [comment]
`modifiedFiles` can be undefined, so I think it is bug in types, not in code

**Comment by sokra** [comment]
`modifiedFiles` is always `undefined` in the initial compilation. We need to fix the typings.

But in your case it actually shouldn't be undefined. We need to fix that too.

**Comment by alan-agius4** [comment]
@sokra , indeed in this case it does need to be defined.

**Comment by alan-agius4** [comment]
Fixed in 5.38.0
